Add a stall detector histogram

amnonh opened the pull request:
Seastar reports stalls above some threshold. It is helpful to follow the  distribution of the stalls. 

This series adds a time_estimated_histogram to the stall detector.
The series takes the estimated_histogram definition from the ScyllaDB project.
    
Whenever a stall is found, it is added to the histogram.
The stall histogram is reported once per node.
    
Fixes #1674

raphaelsc commented on src/core/reactor.cc:
```diff
@@ -1111,7 +1111,9 @@ cpu_stall_detector::cpu_stall_detector(cpu_stall_detector_config cfg)
-            sm::make_counter("reported", _total_reported, sm::description("Total number of reported stalls, look in the traces for the exact reason"))});
+            sm::make_counter("reported", _total_reported, sm::description("Total number of reported stalls, look in the traces for the exact reason")),
+            sm::make_histogram("stalls", sm::description("A histogram stalls"), [this] {return _stalls_histogram.to_metrics_histogram();}).aggregate({seastar::metrics::shard_label}).set_skip_when_empty()
+    });
```
I think we can easily cache seastar::metrics::histogram, and only recompute it once the histogram is updated.

raphaelsc replied:
Updating the histogram will be rare, as it only happens once we have a stall in a shard, therefore, computing the metric histogram will be cheap most of the time.

amnonh replied:
While this is true, it's an unneeded optimization. A calculation is done only when scrapping a metric, typically once every 20-30 seconds.

Reducing that calculation is neglectable.

raphaelsc replied:
A large number of small incidents can add up to a disaster.

Users won't be happy with spikes every 20-30 seconds.

Not claiming this specific metric is the root of all evil, but I think the rule # 1 should be make every metric computation as fast as possible. Expose some metric precomputed. E.g. per-table disk usage. Will I compute the disk usage by iterating through all sstables when scraping the metric, or I will take advantage that I can precompute it once updating the sstable set? The latter.

to_metrics_histogram() allocates some memory, iterates through each bucket, and aggregate some stats from each.

amnonh replied:
but we have one such metric per shard

amnonh replied:
@raphaelsc not sure if you missed something or if  missed something.
once every scrap interval (lets say 20s) the metric layer collect the value.
it takes a copy of the data (it's preamptable so things can change while reporting).
We don't save much if we hold that copy (we're talking about 60*float of data).

We have many more histograms in the system, so saving this one will have zero impact.
In the future I'll be moving to Prometheus native histograms, that's were more optimization would have effect.

michoecho replied:
@raphaelsc Your line of thinking is dangerous, because it leads to a positive feedback loop of performance problems.

You are basically saying: "running this performance diagnostic code might cause performance problems, so let's run it only when performance problems happen". That's the kind of thing that invites a disaster. If there really is a performance problem in diagnostic code, it's better if it happens every 20s (so we can catch it with any performance test and react), rather than have it happen conditionally during other performance problems and amplify them.

That aside, as @amnonh says, outputting a duration histogram to wire every 20s should be nowhere near a performance problem. (And if it is a performance problem, the entire metrics framework deserves to be burned).

raphaelsc replied:
> @raphaelsc Your line of thinking is dangerous, because it leads to a positive feedback loop of performance problems.
> 
> You are basically saying: "running this performance diagnostic code might cause performance problems, so let's run it only when performance problems happen". _That's_ the kind of thing that invites a disaster. If there really is a performance problem in diagnostic code, it's better if it happens unconditionally every 20s (so we can catch it with any performance test and react), rather than have it happen conditionally during other performance problems and amplify them.
> 

Not at all. You're thinking I was suggest hiding the dirt under the carpet, a performance issue that comes from to_metrics_histogram() for example, but that's not what I meant. I was only suggesting that as a rule of thumb, we want a metric implementation to be as fast as possible. Otherwise, we'll be unnecessarily burning cpu cycles which could be devouted elsewhere (like exposing more metrics during the scraping). For example, if I want to expose some metric from sstable set (e.g. compaction backlog), will I iterate through sstables and perform the calculation or will I take advantage of the fact that the metric can be precomputed when updating the set? Exposing a metric is not free, and there's a real overhead that can affect the system. There's a reason why we have real issues with monitoring when enabling per-table metrics.

amnonh replied:
@raphaelsc the problem we have with per-table metrics is because we have histograms per table per shard, that's alot, this is one histogram per node. one.
It's neglectable. There's a future work of making histogram reporting faster, but that would have impact one you're talking about many histogram, for a single histogram it will have no effect

michoecho replied:
> I was only suggesting that as a rule of thumb, we want a metric implementation to be as fast as possible.

@raphaelsc But if I understand correctly, you aren't suggesting to speed it up — you are suggesting to cache it. This will save you some CPU cycles, yes, but only in the case when there are no stalls during the scrape period, which is a condition that we don't control.

This is an opportunistic optimization — one that speeds the system up, but only in lucky/happy conditions.
In a latency-sensitive system, such optimizations are risky in general because they make the behavior of the system harder to predict. But they are extra risky if their conditions depend on performance, because that causes a feedback loop.

In this particular case, we could claim the following: "If there are frequent stalls, the proposed change has no effect. If there are no stalls, the proposed change speeds things up. So the change is a clear improvement".
But the dual of that is: "After the change, when the system has performance problems (which is likely to cause stalls), producing the metric is more expensive than usual, so the system has even more performance problems".

Another example of this trope in Scylla is the row cache. It doesn't have a fixed size, but balloons to fill all unused memory in the system. In the usual case this makes the cache bigger, which makes requests cheaper, which means that the peak throughput of the system is greater.
But the dual of that is: if the system gets overloaded, requests and maintenance work might start to queue up, which will steal memory from the cache, which will make the cache smaller, which will make requests more expensive to serve. And in the end, this design causes the system to respond to overload by *slowing down*.

Do you see what I mean? The work saved by opportunistic optimizations is not truly gone — it's hidden under the carpet, ready to come back when that would be the most problematic.

raphaelsc replied:
> > I was only suggesting that as a rule of thumb, we want a metric implementation to be as fast as possible.
> 
> @raphaelsc But if I understand correctly, you aren't suggesting to speed it up — you are suggesting to cache it. This will save you some CPU cycles, yes, but only in the case when there are no stalls during the scrape period, which is a condition that we don't control.
> 
> This is an opportunistic optimization — one that speeds the system up, but only in lucky/happy conditions. In a latency-sensitive system, such optimizations are risky in general because they make the behavior of the system harder to predict. But they are extra risky if their conditions depend on performance, because that causes a feedback loop.
> 
> In this particular case, we could claim the following: "If there are frequent stalls, the proposed change has no effect. If there are no stalls, the proposed change speeds things up. So the change is a clear improvement". But the dual of that is: "After the change, when the system has performance problems (which is likely to cause stalls), producing the metric is more expensive than usual, so the system has even more performance problems".
> 
> Another example of this trope in Scylla is the row cache. It doesn't have a fixed size, but balloons to fill all unused memory in the system. In the usual case this makes the cache bigger, which makes requests cheaper, which means that the peak throughput of the system is greater. But the dual of that is: if the system gets overloaded, requests and maintenance work might start to queue up, which will steal memory from the cache, which will make the cache smaller, which will make requests more expensive to serve. And in the end, this design causes the system to respond to overload by _slowing down_.
> 
> Do you see what I mean? The work saved by opportunistic optimizations is not truly gone — it's hidden under the carpet, ready to come back when that would be the most problematic.

What you wrote above makes total sense, and I agree completely with you.

avikivity commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+ * Limitation: You must set the MIN value to be higher or equal to the Precision.
+ *
+ */
+template<uint64_t Min, uint64_t Max, size_t Precision>
```
Please add \tparam annotations to explain the parameters directly.

avikivity commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+ *
+ */
+template<uint64_t Min, uint64_t Max, size_t Precision>
+class approx_exponential_histogram {
```
s/approx/approximate/

avikivity replied:
Seastar naming convention is to use full names. It shouldn't be affected by the donor repository.

If ScyllaDB wants to use it, it can use a type synonym, or update the mentions.

avikivity commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+    static constexpr size_t NUM_BUCKETS = NUM_EXP_RANGES * Precision + 1;
+    static constexpr unsigned PRECISION_BITS = log2floor(Precision);
+    static constexpr unsigned BASESHIFT = log2floor(Min);
+    static constexpr uint64_t LOWER_BITS_MASK = Precision - 1;
```
Can these be made private?

avikivity replied:
?

amnonh replied:
They are now private

avikivity commented on include/seastar/core/internal/stall_detector.hh:
```diff
@@ -59,6 +60,7 @@ protected:
     std::atomic<uint64_t> _stall_detector_missed_ticks = { 0 };
     unsigned _reported = 0;
     unsigned _total_reported = 0;
+    metrics::time_estimated_histogram _stalls_histogram;
```
You're not using approx_blah_historgram here.

amnonh replied:
@avikivity right, but time_estimated_histogram uses it.

I've took it as-is from scyllaDB repository, in the future we can decide to keep only this copy

amnonh commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+#include <array>
+
+namespace seastar {
+namespace metrics {
```
it's public for users of the metrics layer

tchaikov replied:
we can use `namespace seastar::metrics { ... }` since C++17, IMHO, it's more readable than the nested notation. see syntax (8) in https://en.cppreference.com/w/cpp/language/namespace.

avikivity replied:
I know it's public, but should it be? Making it public means we can't change it later.

amnonh commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+    static constexpr unsigned BASESHIFT = log2floor(Min);
+    static constexpr uint64_t LOWER_BITS_MASK = Precision - 1;
+private:
+    std::array<uint64_t, NUM_BUCKETS> _buckets;
```
I didn't understnad this comment

tchaikov replied:
nit, could instead initialize `_buckets` using aggregation initialization, like:
```c++
std::array<uint64_t, NUM_BUCKETS> _buckets = {};
```
instead of calling `clear()` in the constructor. IMHO, it'd be more clear that to initialize member variable where they are defined.

avikivity replied:
If you use std::array, you should include \<array> and not rely on it getting here somehow, because some day it won't.

amnonh replied:
it's included

amnonh commented:
@avikivity V2 had the following changes:
const expr are now private
add \param for the the template parameters
add a unit test for the histogram

tchaikov commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+            elements += _buckets[i];
+            sum += _buckets[i] * get_bucket_lower_limit(i);
+        }
+        return (elements) ?  sum / elements : 0;
```
nit, might want to drop the parenthesis  around `elements` .

amnonh commented:
> @avikivity V2 had the following changes: const expr are now private add \param for the the template parameters add a unit test for the histogram

@avikivity can you take a look at V2?

michoecho commented on src/core/reactor.cc:
```diff
@@ -1401,6 +1403,7 @@ void cpu_stall_detector::generate_trace() {
     auto delta = reactor::now() - _run_started_at;
 
     _total_reported++;
+    _stalls_histogram.add(delta);
```
This is the wrong place to put the histogram increment.

 `generate_trace()` is rate limited. We have to ratelimit the logs out of necessity, but it would be undesirable to ratelimit histogram increments.

Also, `cpu_stall_detector` reports every long stall multiple times (as its millisecond duration crosses consecutive power of 2), which would render the histogram confusing at least.

I strongly suggest moving this line to the `if` inside `reactor::account_runtime`:
```
void
reactor::account_runtime(task_queue& tq, sched_clock::duration runtime) {
    if (runtime > (2 * _task_quota)) {
        tq._time_spent_on_task_quota_violations += runtime - _task_quota;
    }
    tq._vruntime += tq.to_vruntime(runtime);
    tq._runtime += runtime;
}
```

This way we both avoid the two problems I mentioned above (rate limiting and accounting the same stall multiple times), and we make the histogram count all stalls (as defined by `task_quota_ms`, not just the *logged* "stalls" as defined by `blocked-reactor-notify-ms`, which is much more useful.

amnonh replied:
@michoecho thanks for the suggestion, I just forced push v3 with the changes

michoecho reviewed (changes requested):
The PR looks good to me, and it implements a desirable feature.

But I have a change request, because I think you picked the wrong place to increment the histogram. You increment it in `generate_trace()`, which is called when stalls are *logged*. But not all stalls are logged, and some stalls are logged multiple times. It would be much better to move this histogram increment to the source — the reactor — and account all stalls, not only the logged ones.

avikivity commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+inline time_estimated_histogram time_estimated_histogram_merge(time_estimated_histogram a, const time_estimated_histogram& b) {
+    return a.merge(b);
+}
+
```
?

amnonh replied:
I've added unit test

avikivity commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+ * 512us, 640us, 768us, 896us, 1024us, 1280us, 1536us, 1792us...16s, 20s, 25s, 29s, 33s (33554432us)
+ */
+class time_estimated_histogram : public approx_exponential_histogram<512, 33554432, 4> {
+public:
```
Why is time_estimated_histogram choosing these values? Different measurements will need different parameters. Best to keep this a template.

amnonh replied:
It's taken from Scylla As-Is, I took it with as minimal changes as possible, I can link to the discussion when it was added back then.

When I'll do more changes to the histograms mechanism, I'll update it here, but i think we should be consisten between the implementations.

michoecho replied:
@avikivity Are you aware that all histogram code in this PR was copied as-is from the Scylla repository? Focusing the review on years-old code, which @amnonh isn't an author of, seems counterproductive.

avikivity replied:
Taking 7 year old mistakes and enshrining them as an API is not a way to bypass review.

We bypassed review in another way, by making things internal, but copy/paste in general is not an excuse for bad code.

avikivity commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+    }
+
+    void add_micro(uint64_t n) {
+        approx_exponential_histogram<512, 33554432, 4>::add(n);
```
```
template <...>
add(std::chrono::duration<>
```

amnonh replied:
I can template the add(duration)

michoecho commented on tests/unit/metrics_test.cc:
```diff
@@ -319,3 +320,57 @@ SEASTAR_THREAD_TEST_CASE(test_relabel_enable_disable_skip_when_empty) {
+    using namespace std::chrono_literals;
+    time_estimated_histogram histogram1;
+    time_estimated_histogram histogram2;
+    time_estimated_histogram::duration min = std::chrono::microseconds(512);
```
nit: magic number. It should have a comment, like:
```
// The lower bound of time_estimated_histogram is 512 us.
time_estimated_histogram::duration min = std::chrono::microseconds(512);
```

amnonh replied:
It's a test. Those are arbitrary numbers. I chose the same numbers I used for time_estimated_histogram, but it could have been anything (as long as it's  power of 2)

michoecho replied:
> Those are arbitrary numbers. I chose the same numbers I used for time_estimated_histogram, but it could have been anything (as long as it's power of 2)

What do you mean? If you changed the 512 here to any other number, the test would fail. Likewise, if any constant in time_estimated_histogram changed, the test would fail.

michoecho commented on tests/unit/metrics_test.cc:
```diff
@@ -319,3 +320,57 @@ SEASTAR_THREAD_TEST_CASE(test_relabel_enable_disable_skip_when_empty) {
+    time_estimated_histogram::duration next = min*2;
+
+    for (size_t i = 0; i < 16; i++) {
+        auto delta = (next - min)/4;
```
nit: magic number. It would be better to have something like:
```
// The number of linearly-spaced buckets between consecutive powers of 2 in time_estimated_histogram.
constexpr int PRECISION = 4;
...
auto delta = (next - min) / PRECISION;
```

michoecho commented on src/core/reactor.cc:
```diff
@@ -2519,6 +2520,7 @@ void reactor::register_metrics() {
             sm::make_total_bytes("aio_bytes_write", _io_stats.aio_write_bytes, sm::description("Total aio-writes bytes")),
             sm::make_counter("aio_outsizes", _io_stats.aio_outsizes, sm::description("Total number of aio operations that exceed IO limit")),
             sm::make_counter("aio_errors", _io_stats.aio_errors, sm::description("Total aio errors")),
+            sm::make_histogram("stalls", sm::description("A histogram of reactor stalls duration"), [this] {return _stalls_histogram.to_metrics_histogram();}).aggregate({seastar::metrics::shard_label}).set_skip_when_empty(),
```
typo: reactor stalls duration -> reactor stall durations

michoecho reviewed (changes requested):
Looks good to me, but I think we should make the histograms private/internal (by moving them to different folders and or to the `internal` namespace), because we don't want them used by users of Seastar.

I also added a few very minor nitpicks.

I didn't review the histogram code, because I think that's out of scope here.

amnonh commented:
> Looks good to me, but I think we should make the histograms private/internal (by moving them to different folders and or to the `internal` namespace), because we don't want them used by users of Seastar.
> 

Yes, following @avikivity comment I've moved them into impl name space, compiling and testing it now

amnonh commented:
@avikivity @michoecho 
I've forced push V4
rename to approximate_exponential_histogram
histograms are now under impl namespace.
I've used a template member function for adding a duration
In the unit test, I've replaced numbers with consexpr
fixed the typo in the metric definition.

amnonh commented:
@avikivity ping

avikivity commented on include/seastar/core/estimated_histogram.hh:
```diff
@@ -0,0 +1,385 @@
+
+namespace seastar {
+namespace metrics {
+namespace impl {
```
Seastar uses `internal` for internal stuff.

You can write `namespace seastar::metrics::internal {` instead of nesting three statements.

Please move the file to core/internal so it's not used by users.

amnonh commented:
@avikivity I've moved to internal and pull rebased and forced push an updated version
